test/e2e/vsphere_volume_placement.go

```diff
@@ -0,0 +1,276 @@
+		It("should provision pod on the node with matching label", func() {
+			By("creating vmdk")
+			vsp, err := vsphere.GetVSphere()
+			Expect(err).NotTo(HaveOccurred())
+
+			volumePath, err = createVSphereVolume(vsp, nil)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Creating pod on the node: " + node1Name)
+			pod := getPodSpec(volumePath, node1KeyValueLabel, nil)
+
+			pod, err = c.CoreV1().Pods(ns).Create(pod)
+			Expect(err).NotTo(HaveOccurred())
+			By("Waiting for pod to be ready")
+			Expect(f.WaitForPodRunning(pod.Name)).To(Succeed())
+
+			By("Verify volume is attached to the node: " + node1Name)
+			isAttached, err := verifyVSphereDiskAttached(vsp, volumePath, types.NodeName(node1Name))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(isAttached).To(BeTrue(), "disk is not attached with the node")
+
+			By("Deleting pod")
+			err = c.CoreV1().Pods(ns).Delete(pod.Name, nil)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Waiting for volume to be detached from the node")
+			waitForVSphereDiskToDetach(vsp, volumePath, types.NodeName(node1Name))
+
+			By("Creating pod on the same node: " + node1Name)
+			pod = getPodSpec(volumePath, node1KeyValueLabel, nil)
+			pod, err = c.CoreV1().Pods(ns).Create(pod)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Waiting for pod to be running")
+			Expect(f.WaitForPodRunning(pod.Name)).To(Succeed())
+
+			By("Verify volume is attached to the node: " + node1Name)
+			isAttached, err = verifyVSphereDiskAttached(vsp, volumePath, types.NodeName(node1Name))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(isAttached).To(BeTrue(), "disk is not attached with the node")
+
+			By("Deleting pod")
+			err = c.CoreV1().Pods(ns).Delete(pod.Name, nil)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Waiting for volume to be detached from the node: " + node1Name)
+			waitForVSphereDiskToDetach(vsp, volumePath, types.NodeName(node1Name))
+
+			By("Deleting vmdk")
+			if len(volumePath) > 0 {
+				vsp.DeleteVolume(volumePath)
+			}
+		})
+	})
+
+	/*
+		Test Steps
+		-----------
+		1. Find two nodes with the status available and ready for scheduling.
+		2. Add labels to the both nodes. - (vsphere_e2e_label: Random UUID)
+		3. Create VMDK volume
+		4. Create POD Spec with volume path of the vmdk and NodeSelector set to node1's label.
+		5. Create POD and wait for POD to become ready.
+		6. Verify volume is attached to the node.
+		7. Delete POD.
+		8. Wait for volume to be detached from the node1.
+		9. Create POD Spec with volume path of the vmdk and NodeSelector set to node2's label.
+		10. Create POD and wait for POD to become ready.
+		11. Verify volume is attached to the node2.
+		12. Delete POD.
+		13. Delete VMDK volume.
+	*/
+	framework.KubeDescribe("Test Back-to-back pod creation/deletion with the same volume source attach/detach to different worker nodes", func() {
+		It("should provision pod on the node with matching label", func() {
```

Comment:
added following helper funcs to avoid duplication.

createPodWithVolumeAndNodeSelector(client clientset.Interface, namespace string, vsp *vsphere.VSphere, nodeName string, nodeKeyValueLabel map[string]string, volumePath string) (pod *v1.Pod)

deletePodAndWaitForVolumeToDetach(client clientset.Interface, namespace string, vsp *vsphere.VSphere, nodeName string, pod *v1.Pod, volumePath string)

each It blocks is now reduced to 5 lines.